File: lilac/tink/service.go
package tink

import (
	"context"
	"time"

	"google.golang.org/protobuf/types/known/emptypb"

	crypto "github.com/saturn-xiv/palm/lilac/env/crypto"
	pb "github.com/saturn-xiv/palm/lilac/tink/v2"
)

type JwtService struct {
	pb.UnimplementedJwtServer

	client *crypto.Jwt
}

func NewJwtService(jwt *crypto.Jwt) *JwtService {
	return &JwtService{client: jwt}
}

func (p JwtService) Sign(_ctx context.Context, req *pb.JwtSignRequest) (*pb.JwtSignResponse, error) {
	expiry := time.Second * time.Duration(req.Ttl.Seconds)
	token, err := p.client.Sign(req.Issuer, req.Subject, req.Audience, req.Claims, expiry)
	if err != nil {
		return nil, err
	}
	return &pb.JwtSignResponse{Token: token}, nil
}

func (p JwtService) Verify(_ctx context.Context, req *pb.JwtVerifyRequest) (*pb.JwtVerifyResponse, error) {
	jwt_id, subject, claims, err := p.client.Verify(req.Token, req.Issuer, req.Audience)
	if err != nil {
		return nil, err
	}

	return &pb.JwtVerifyResponse{
		Id:      jwt_id,
		Subject: subject,
		Claims:  claims,
	}, nil
}

type HmacService struct {
	pb.UnimplementedHmacServer

	client *crypto.HMac
}

func NewHmacService(hmac *crypto.HMac) *HmacService {
	return &HmacService{client: hmac}
}

func (p HmacService) Sign(_ctx context.Context, req *pb.HmacSignRequest) (*pb.HmacSignResponse, error) {
	code, err := p.client.Sign(req.Plain)
	if err != nil {
		return nil, err
	}
	return &pb.HmacSignResponse{Code: code}, nil
}

func (p HmacService) Verify(_ctx context.Context, req *pb.HmacVerifyRequest) (*emptypb.Empty, error) {
	if err := p.client.Verify(req.Code, req.Plain); err != nil {
		return nil, err
	}
	return &emptypb.Empty{}, nil
}

type AesService struct {
	pb.UnimplementedAesServer

	client *crypto.Aes
}

func NewAesService(aes *crypto.Aes) *AesService {
	return &AesService{client: aes}
}

func (p AesService) Encrypt(_ctx context.Context, req *pb.PlainAesMessage) (*pb.CodeAesMessage, error) {
	code, salt, err := p.client.Encrypt(req.Plain)
	if err != nil {
		return nil, err
	}
	return &pb.CodeAesMessage{Code: code, Salt: salt}, nil
}

func (p AesService) Decrypt(_ctx context.Context, req *pb.CodeAesMessage) (*pb.PlainAesMessage, error) {
	plain, err := p.client.Decrypt(req.Code, req.Salt)
	if err != nil {
		return nil, err
	}
	return &pb.PlainAesMessage{Plain: plain}, nil
}

File: loquat/src/service.cpp
#include "loquat/service.hpp"
#include "loquat/env.hpp"

grpc::Status loquat::JwtService::Sign(
    grpc::ServerContext* context,
    const palm::loquat::v1::JwtSignRequest* request,
    palm::loquat::v1::JwtSignResponse* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  loquat::Jwt jwt;

  const std::optional<std::string> jwt_id =
      request->has_jwt_id() ? std::optional<std::string>{request->jwt_id()}
                            : std::nullopt;
  const std::optional<std::string> key_id =
      request->has_key_id() ? std::optional<std::string>{request->key_id()}
                            : std::nullopt;
  const std::optional<std::string> payload =
      request->has_payload() ? std::optional<std::string>{request->payload()}
                             : std::nullopt;

  std::set<std::string> audiences;
  for (const auto& it : request->audiences()) {
    audiences.insert(it);
  }
  const auto token =
      jwt.sign(jwt_id, key_id, request->issuer(), request->subject(), audiences,
               absl::FromUnixSeconds(request->issued_at()),
               absl::FromUnixSeconds(request->not_before()),
               absl::FromUnixSeconds(request->expired_at()), payload);
  if (token) {
    response->set_token(token.value());
    return grpc::Status::OK;
  }
  return grpc::Status(grpc::StatusCode::INTERNAL, "Jwt sign failed");
}
grpc::Status loquat::JwtService::Verify(
    grpc::ServerContext* context,
    const palm::loquat::v1::JwtVerifyRequest* request,
    palm::loquat::v1::JwtVerifyResponse* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  loquat::Jwt jwt;
  const auto it =
      jwt.verify(request->token(), request->issuer(), request->audience());
  if (it) {
    response->set_subject(it->first);
    if (it->second) {
      response->set_payload(it->second.value());
    }
    return grpc::Status::OK;
  }
  return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                      "JWT verification check failed");
}
grpc::Status loquat::HMacService::Sign(
    grpc::ServerContext* context,
    const palm::loquat::v1::HMacSignRequest* request,
    palm::loquat::v1::HMacSignResponse* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  loquat::HMac mac;
  const auto hashed = mac.sign(request->plain());
  if (hashed) {
    response->set_hashed(hashed.value());
    return grpc::Status::OK;
  }
  return grpc::Status(grpc::StatusCode::INTERNAL, "HMac sign failed");
}
grpc::Status loquat::HMacService::Verify(
    grpc::ServerContext* context,
    const palm::loquat::v1::HMacVerifyRequest* request,
    palm::loquat::v1::Empty* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  loquat::HMac mac;
  return mac.verify(request->hashed(), request->plain())
             ? grpc::Status::OK
             : grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                            "HMac verification check failed");
  ;
}
grpc::Status loquat::AesService::Encrypt(
    grpc::ServerContext* context,
    const palm::loquat::v1::AesEncryptRequest* request,
    palm::loquat::v1::AesEncryptResponse* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  loquat::Aes aes;
  const auto cipher = aes.encrypt(request->plain(), request->associated_data());
  if (cipher) {
    response->set_cipher(cipher.value());
    return grpc::Status::OK;
  }
  return grpc::Status(grpc::StatusCode::INTERNAL, "AES encrypt failed");
}
grpc::Status loquat::AesService::Decrypt(
    grpc::ServerContext* context,
    const palm::loquat::v1::AesDecryptRequest* request,
    palm::loquat::v1::AesDecryptResponse* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  loquat::Aes aes;
  const auto plain = aes.decrypt(request->cipher(), request->associated_data());
  if (plain) {
    response->set_plain(plain.value());
    return grpc::Status::OK;
  }
  return grpc::Status(grpc::StatusCode::INTERNAL, "AES decrypt failed");
}
grpc::Status loquat::Argon2Service::Sign(
    grpc::ServerContext* context,
    const palm::loquat::v1::Argon2SignRequest* request,
    palm::loquat::v1::Argon2SignResponse* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);
  const auto it = loquat::argon2id::sign(request->password());
  if (it) {
    response->set_hashed(it.value());
    return grpc::Status::OK;
  }
  return grpc::Status(grpc::StatusCode::INTERNAL, "sign check failed");
}
grpc::Status loquat::Argon2Service::Verify(
    grpc::ServerContext* context,
    const palm::loquat::v1::Argon2VerifyRequest* request,
    palm::loquat::v1::Empty* response) {
  spdlog::debug("call {}", __PRETTY_FUNCTION__);

  return loquat::argon2id::verify(request->hashed(), request->password())
             ? grpc::Status::OK
             : grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                            "verification check failed");
}
